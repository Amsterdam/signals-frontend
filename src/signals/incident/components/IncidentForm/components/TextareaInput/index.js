import React from 'react';
import PropTypes from 'prop-types';

import Header from '../Header/';

const TextareaInput = ({ handler, touched, value, hasError, meta, parent, getError, validatorsOrOpts }) => (
  <div className={`${meta && meta.isVisible ? 'row' : ''}`}>
    {meta && meta.isVisible ?
      <div className={`${meta.className || 'col-12'} mode_input`}>
        <Header
          meta={meta}
          options={validatorsOrOpts}
          touched={touched}
          hasError={hasError}
          getError={getError}
        >
          <div className="invoer">
            <textarea
              placeholder={meta.placeholder}
              {...handler()}
              onBlur={(e) => parent.meta.updateIncident({ [meta.name]: e.target.value })}
            />
            { meta.maxLength &&
              <div className="input-help">
                <span className="text-area-input__counter">
                  {`${value ? value.length : '0'}/${meta.maxLength} tekens` }
                </span>
              </div>
            }
          </div>
        </Header>
      </div>
       : ''}
  </div>
);

TextareaInput.propTypes = {
  handler: PropTypes.func,
  touched: PropTypes.bool,
  value: PropTypes.string,
  hasError: PropTypes.func,
  meta: PropTypes.object,
  parent: PropTypes.object,
  getError: PropTypes.func,
  validatorsOrOpts: PropTypes.object
};

export default TextareaInput;
