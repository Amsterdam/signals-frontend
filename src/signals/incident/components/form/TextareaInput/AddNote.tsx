import { useState, useCallback, useRef } from 'react'
import type { FC, FocusEvent } from 'react'

import GenericAddNote, { getAddNoteError } from 'components/AddNote'
import type { AddNoteProps } from 'components/AddNote/AddNote'

const AddNote: FC<AddNoteProps> = ({ maxContentLength, onBlur, ...props }) => {
  const [error, setError] = useState('')
  const textAreaRef = useRef<HTMLTextAreaElement>(null)

  const onNoteBlur = useCallback(
    (event: FocusEvent<HTMLTextAreaElement>) => {
      const text = event.target.value.trim()

      if (!text) return

      if (maxContentLength) {
        const validationError = getAddNoteError({
          maxContentLength,
          text,
          shouldContainAtLeastOneChar: false,
        })

        if (validationError !== '') {
          setError(validationError)
          return
        }
      }

      onBlur && onBlur(event)
    },
    [maxContentLength, onBlur]
  )

  const onChange = useCallback((event) => {
    const { value } = event.target

    if (value.trim() !== '') {
      setError('')
    }
  }, [])

  return (
    <GenericAddNote
      {...props}
      error={error}
      maxContentLength={maxContentLength}
      onChange={onChange}
      onBlur={onNoteBlur}
      ref={textAreaRef}
      isStandalone={false}
      rows={5}
    />
  )
}

export default AddNote
