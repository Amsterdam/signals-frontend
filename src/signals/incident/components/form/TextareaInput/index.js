import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import TextArea from 'components/TextArea';

import Header from '../Header';

const TextareaInput = ({ handler, touched, value, hasError, meta, parent, getError, validatorsOrOpts }) =>
  meta?.isVisible && (
    <Header meta={meta} options={validatorsOrOpts} touched={touched} hasError={hasError} getError={getError}>
      <Fragment>
        <TextArea
          placeholder={meta.placeholder}
          {...handler()}
          onBlur={e =>
            parent.meta.updateIncident({
              [meta.name]: meta.autoRemove ? e.target.value.replace(meta.autoRemove, '') : e.target.value,
            })
          }
          helpText={meta.maxLength > 0 && `${value ? value.length : '0'}/${meta.maxLength} tekens`}
        />
      </Fragment>
    </Header>
  );

TextareaInput.propTypes = {
  handler: PropTypes.func,
  touched: PropTypes.bool,
  value: PropTypes.string,
  hasError: PropTypes.func,
  meta: PropTypes.object,
  parent: PropTypes.object,
  getError: PropTypes.func,
  validatorsOrOpts: PropTypes.object,
};

export default TextareaInput;
