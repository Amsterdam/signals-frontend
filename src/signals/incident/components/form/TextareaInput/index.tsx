import TextareaInput from './TextareaInput';

export default TextareaInput;
