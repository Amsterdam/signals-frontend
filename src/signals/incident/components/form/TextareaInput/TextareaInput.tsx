// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2022 Gemeente Amsterdam
import type { FocusEvent, FunctionComponent } from 'react'

import styled from 'styled-components'

import FormField from 'components/FormField'
import type { FormInputProps } from 'types/reactive-form'

import AddNote from './AddNote'

export type TextAreaInputProps = Omit<FormInputProps, 'handler'>

const ThinLabel = styled.span`
  font-weight: 400;
`

const TextareaInput: FunctionComponent<TextAreaInputProps> = ({
  hasError,
  getError,
  meta,
  parent,
  validatorsOrOpts,
  value,
}) => {
  if (!meta) return null

  const { label, subtitle, name, isVisible, maxLength } = meta

  if (!isVisible || !name) return null

  const onBlur = (event: FocusEvent<HTMLTextAreaElement>) => {
    const text = event.target.value

    parent.meta.updateIncident({
      [name]: text,
    })
  }

  const props = {
    isStandalone: false,
    name,
    subtitle,
    maxContentLength: maxLength || undefined,
    label: (
      <>
        {label} <ThinLabel>(niet verplicht)</ThinLabel>
      </>
    ),
    onBlur,
    value,
  }

  return (
    <FormField
      meta={meta}
      options={validatorsOrOpts}
      hasError={hasError}
      getError={getError}
    >
      <AddNote inForm id={meta.name} {...props} />
    </FormField>
  )
}

export default TextareaInput
