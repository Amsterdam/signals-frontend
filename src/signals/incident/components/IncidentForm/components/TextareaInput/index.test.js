import React from 'react';
import { shallow } from 'enzyme';

import TextareaInput from './index';

describe('Form component <TextareaInput />', () => {
  const metaFields = {
    name: 'input-field-name',
    placeholder: 'type here'
  };
  let wrapper;
  let handler;
  let touched;
  let getError;
  let hasError;
  let parent;

  beforeEach(() => {
    handler = jest.fn();
    touched = false;
    getError = jest.fn();
    hasError = jest.fn();
    parent = {
      meta: {
        setIncident: jest.fn()
      }
    };

    wrapper = shallow(<TextareaInput
      handler={handler}
      parent={parent}
      touched={touched}
      hasError={hasError}
      getError={getError}
    />);
  });

  describe('rendering', () => {
    it('should render text area field correctly', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true
        }
      });

      expect(handler).toHaveBeenCalledWith();
      expect(wrapper).toMatchSnapshot();
    });

    it('should render no text area field when not visible', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: false
        }
      });

      expect(handler).not.toHaveBeenCalled();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('events', () => {
    const event = { target: { value: 'diabolo' } };

    it('sets incident when value changes', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true,
          updateIncident: true
        }
      });

      wrapper.find('textarea').simulate('change', event);

      expect(parent.meta.setIncident).toHaveBeenCalledWith({
        'input-field-name': 'diabolo'
      });
    });

    it('does nothing when updateIncident is false', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true,
          updateIncident: false
        }
      });

      wrapper.find('textarea').simulate('change', event);

      expect(parent.meta.setIncident).not.toHaveBeenCalled();
    });
  });
});
