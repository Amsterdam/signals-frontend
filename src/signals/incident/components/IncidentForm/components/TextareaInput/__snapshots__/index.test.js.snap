// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component <TextareaInput /> rendering should render no text area field when not visible 1`] = `
<div
  className=""
/>
`;

exports[`Form component <TextareaInput /> rendering should render text area field correctly 1`] = `
<div
  className="row"
>
  <div
    className="col-12 mode_input"
  >
    <Header
      getError={[MockFunction]}
      hasError={[MockFunction]}
      meta={
        Object {
          "isVisible": true,
          "name": "input-field-name",
          "placeholder": "type here",
        }
      }
      touched={false}
    >
      <div
        className="invoer"
      >
        <textarea
          onChange={[Function]}
          placeholder="type here"
        />
      </div>
    </Header>
  </div>
</div>
`;
